node/circuits.js

```diff
@@ -0,0 +1,238 @@
+}
+
+Circuits.prototype.getCircuit = function getCircuit(callerName, serviceName, endpointName) {
+    var self = this;
+    var circuits = self.circuitsByServiceName['$' + serviceName];
+    if (!circuits) {
+        circuits = new ServiceCircuits(self);
+        self.circuitsByServiceName['$' + serviceName] = circuits;
+    }
+    return circuits.getCircuit(callerName, serviceName, endpointName);
+};
+
+Circuits.prototype.getCircuitTuples = function getCircuitTuples() {
+    var self = this;
+    var circuits = Object.keys(self.circuitsByServiceName).map(function (serviceName) {
+        return self.circuitsByServiceName[serviceName].getCircuitTuples();
+    });
+    return Array.prototype.concat.apply([], circuits);
+};
+
+Circuits.prototype.handleRequest = function handleRequest(req, buildRes, nextHandler) {
+    var self = this;
+    // Default the caller name.
+    // All callers that fail to specifiy a cn share a circuit for each sn:en
+    // and fail together.
+    var callerName = req.headers.cn || 'yunocn';
+    var serviceName = req.serviceName;
+    if (!serviceName) {
+        return buildRes().sendError('BadRequest', 'All requests must have a service name');
+    }
+    return req.withArg1(function withArg1() {
+        var circuit = self.getCircuit(callerName, serviceName, String(req.arg1));
+        return circuit.handleRequest(req, buildRes, nextHandler);
+    });
+};
+
+// Called upon membership change to collect services that the corresponding
+// exit node is no longer responsible for.
+Circuits.prototype.updateServices = function updateServices() {
+    var self = this;
+    var serviceNames = Object.keys(self.circuitsByServiceName);
+    for (var index = 0; index < serviceNames.length; index++) {
+        var serviceName = serviceNames[index];
+        if (!self.egressNodes.isExitFor(serviceName)) {
+            delete self.circuitsByServiceName[serviceName];
+        }
+    }
+};
+
+function Circuit() {
+    var self = this;
+    self.circuitName = null;
+    self.callerName = null;
+    self.serviceName = null;
+    self.endpointName = null;
+    self.shouldRequestOptions = null;
+    self.stateOptions = null;
+    StateMachine.call(self);
+    EventEmitter.call(self);
+    self.stateChangedEvent = self.defineEvent('stateChanged');
+}
+
+inherits(Circuit, EventEmitter);
+
+Circuit.prototype.setState = StateMachine.prototype.setState;
+
+Circuit.prototype.handleRequest = function handleRequest(req, buildRes, nextHandler) {
+    var self = this;
+    if (self.state.shouldRequest(req, self.shouldRequestOptions)) {
+        return self.monitorRequest(req, buildRes, nextHandler);
+    } else {
+        return buildRes().sendError('Declined', 'Service is not healthy');
+    }
+};
+
+Circuit.prototype.monitorRequest = function monitorRequest(req, buildRes, nextHandler) {
+    var self = this;
+
+    self.state.onRequest(req);
+
+    req.errorEvent.on(onError);
+
+    function monitorBuildRes(options) {
```

Comment:
that's a big function, any utility to pulling out to a method with a one-line binding closure here?